Sanc 27 implement trip updating api endpoints

Lujarios opened the pull request:
https://codethechangeyyc.atlassian.net/browse/SANC-27?focusedCommentId=10884

API endpoints in api/routers/bookings
- `create`
- `getById`
- `getAll`
- `update`
- `cancel`

Added debug page under `http://localhost:3000/debug/bookings`

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Booking debug UI to create, view, update, change status, cancel, and list bookings.
  * Booking records now require start/end times and include a new "cancelled" status.

* **Chores**
  * Database schema and migrations updated (including passenger_info column rename and full migration SQL).
  * Updated UI libraries and developer tooling versions.

* **Documentation**
  * Added "Initializing an Admin Account" instructions to README (duplicated section).

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Lujarios committed:
SANC 27 - Added bookings routing and test page

Lujarios committed:
SANC 27 - Added bookings routing and test page

Lujarios committed:
SANC 27 - Booking routing fix (made public)

Lujarios committed:
SANC 27 - Booking routing fix (made public)

Lujarios committed:
SANC 27 - getAll for bookings

Lujarios committed:
SANC 27 - getAll for bookings

Lujarios committed:
SANC 27 - Fixed mispelling of cancelled

Lujarios committed:
SANC 27 - Fixed mispelling of cancelled

Lujarios committed:
SANC 27 - Added create route for bookings

Lujarios committed:
SANC 27 - Added create route for bookings

Lujarios committed:
SANC 27 - Fix tripRouter import in root

Lujarios commented:
There was some issue with the `tripRouter` not being imported correctly even though I did not make this. Somehow I pulled this error from main when rebasing.

Lujarios committed:
SANC 27 - Added Create Booking to bookings debug page

Lujarios committed:
SANC 27 - Added Create Booking to bookings debug page

burtonjong commented on src/app/debug/bookings/page.tsx:
```diff
@@ -0,0 +1,286 @@
+              borderRadius: "8px",
+              overflowX: "auto",
+              marginTop: "0.75rem",
+            }}
```
all inline styles should be in a module.scss file.

burtonjong commented on src/app/debug/bookings/page.tsx:
```diff
@@ -0,0 +1,286 @@
+        <select
+          value={status}
+          onChange={(e) =>
+            setStatus(e.target.value as "completed" | "incomplete" | "in-progress" | "cancelled")
```
this needs to be in a type or enumeration.

burtonjong replied:
https://github.com/Code-the-Change-YYC/salvationarmy/pull/17/files#diff-11e9171ff36dfa6e6b396909ac6b924cf79337d5f05a7eaa15f78a9b2e2eded6R54-R57

in this pr max has it defined in `types.ts` for now you can make something like this

burtonjong commented on src/app/debug/bookings/page.tsx:
```diff
@@ -0,0 +1,286 @@
+        agencyId: "",
+        driverId: "",
+      });
+    },
+  });
+
+  // Mutations: on success, refresh both single + list
+  const updateMutation = api.bookings.update.useMutation({
+    onSuccess: async () => {
+      await Promise.all([
+        utils.bookings.getById.invalidate({ id: bookingId }),
+        utils.bookings.getAll.invalidate(),
+      ]);
+      bookingQuery.refetch();
+      allBookingsQuery.refetch();
+    },
+  });
```
should be an onfailure handle rhere as well

burtonjong commented on src/app/debug/bookings/page.tsx:
```diff
@@ -0,0 +1,286 @@
+    passengerInfo: "",
+    agencyId: "",
+    driverId: "",
+  });
```
you can use mantine form here pls. Need some frontend validation on some of these fields as well, like the agencyid shouldn't be null, etc.

burtonjong commented on src/server/api/routers/bookings.ts:
```diff
@@ -0,0 +1,101 @@
+
+export const bookingsRouter = createTRPCRouter({
+  // POST /bookings (create)
+  create: publicProcedure
```
this shouldn't be a public procedure

burtonjong commented on src/server/api/routers/bookings.ts:
```diff
@@ -0,0 +1,101 @@
+  }),
+
+  // PATCH /bookings/:id
+  update: publicProcedure
```
not a public procedure either

burtonjong commented on src/server/api/routers/bookings.ts:
none of these should be public procedures, they should be at least logged in

burtonjong commented on src/server/api/routers/bookings.ts:
```diff
@@ -0,0 +1,101 @@
+          title: input.title,
+          pickupLocation: input.pickupLocation,
+          dropoffLocation: input.dropoffLocation,
+          purpose: input.purpose ?? null,
```
dont need ?? operator just do purpose?: input.purpose and for the rest as well

burtonjong commented:
@Lujarios You'll have to rebase/merge main. Please take the incoming changes for the booking schema.

Lujarios committed:
SANC 27 - Made routes private

Lujarios committed:
SANC 27 - Booking page updates to follow project standards

Lujarios committed:
SANC 27 - Fixing pull conficts

Lujarios committed:
Merge branch 'main' into SANC-27-implement-trip-updating-api-endpoints

Lujarios commented:
I haven't yet been able to make a working Auth account on my end to fully test this. I will follow the readme @burtonjong made in tomorrow to get that working locally.

Lujarios commented on src/server/db/booking-schema.ts:
```diff
@@ -2,16 +2,17 @@ import { relations } from "drizzle-orm";
-  status: text("status", { enum: ["incomplete", "completed", "in-progress"] })
-    .notNull()
-    .default("incomplete"),
+  passengerInfo: text("passengerInfo").notNull(),
```
Good catch

Lujarios committed:
SANC-27 — Address CodeRabbit security concerns and update API endpoints

burtonjong commented on src/server/api/routers/bookings.ts:
```diff
@@ -0,0 +1,193 @@
+        .update(bookings)
+        .set({
+          ...updatesToApply,
+          updatedBy: userId,
+        })
+        .where(eq(bookings.id, id))
+        .returning();
+
+      return res[0];
+    }),
+
+  // PATCH /bookings/:id/cancel
+  cancel: protectedProcedure
+    .input(z.object({ id: z.number() }))
+    .mutation(async ({ ctx, input }) => {
+      const existing = await ctx.db
+        .select()
+        .from(bookings)
+        .where(eq(bookings.id, input.id))
+        .limit(1)
+        .then((r) => r[0]);
+
+      if (!existing) {
+        throw new TRPCError({ code: "NOT_FOUND", message: "Booking not found" });
+      }
+
+      return ctx.db
+        .update(bookings)
+        .set({ status: "cancelled", updatedBy: ctx.session.user.id })
+        .where(eq(bookings.id, input.id))
+        .returning()
+        .then((r) => r[0]);
+    }),
```
@Lujarios This one is pretty important - we only want admins to be able to cancel the booking, OR the agency user that made the booking. The one above isn't perfect, but base it off what i put here

Lujarios committed:
Update README with admin account setup instructions

Added instructions for initializing an admin account and using the seed script.

Lujarios committed:
SANC 27 - Added authorization check for cancel booking

Lujarios committed:
Update database migration instructions in README

Clarified instructions for applying database migrations.

Lujarios committed:
SANC 27 - endtime validation

Lujarios committed:
Merge branch 'SANC-27-implement-trip-updating-api-endpoints' of https://github.com/Code-the-Change-YYC/salvationarmy into SANC-27-implement-trip-updating-api-endpoints

Lujarios committed:
SANC 27 - Added admin only to create booking

Lujarios commented on src/server/api/routers/bookings.ts:
```diff
@@ -0,0 +1,216 @@
+    .input(z.object({ id: z.number() }))
+    .mutation(async ({ ctx, input }) => {
+      const existing = await ctx.db
+        .select()
+        .from(bookings)
+        .where(eq(bookings.id, input.id))
+        .limit(1)
+        .then((r) => r[0]);
+
+      if (!existing) {
+        throw new TRPCError({ code: "NOT_FOUND", message: "Booking not found" });
+      }
+
+      const userId = ctx.session.user.id;
+      const role = ctx.session.user.role ?? "user";
+
+      const allowed =
+        role === "admin" ||
+        existing.createdBy === userId ||
+        existing.agencyId === userId ||
+        existing.driverId === userId;
+
+      if (!allowed) {
+        throw new TRPCError({ code: "FORBIDDEN", message: "You cannot cancel this booking." });
+      }
+
+      return ctx.db
+        .update(bookings)
+        .set({ status: "cancelled", updatedBy: ctx.session.user.id })
+        .where(eq(bookings.id, input.id))
+        .returning()
+        .then((r) => r[0]);
+    }),
```
I'll fix this rn.

Lujarios committed:
SANC 27 - Access to data security update

burtonjong commented:
@Lujarios rerequest my review

burtonjong commented:
@Lujarios Feel free to merge now
